lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FlatColumnReader.java

```diff
@@ -0,0 +1,349 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader.flat;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.Slice;
+import io.trino.parquet.DataPage;
+import io.trino.parquet.DataPageV1;
+import io.trino.parquet.DataPageV2;
+import io.trino.parquet.DictionaryPage;
+import io.trino.parquet.ParquetEncoding;
+import io.trino.parquet.PrimitiveField;
+import io.trino.parquet.dictionary.Dictionary;
+import io.trino.parquet.reader.ColumnChunk;
+import io.trino.parquet.reader.ColumnReader;
+import io.trino.parquet.reader.FilteredRowRanges;
+import io.trino.parquet.reader.PageReader;
+import io.trino.parquet.reader.SimpleSliceInputStream;
+import io.trino.parquet.reader.decoders.ValueDecoder;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import org.apache.parquet.io.ParquetDecodingException;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.parquet.ParquetEncoding.RLE;
+import static io.trino.parquet.ParquetReaderUtils.castToByteNegate;
+import static io.trino.parquet.reader.decoders.ValueDecoder.ValueDecodersProvider;
+import static io.trino.parquet.reader.flat.RowRangesIterator.createRowRangesIterator;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public abstract class FlatColumnReader<BufferType>
```

Comment:
Use composition instead of inheritance for code reuse. Here's an alternative approach I prototyped for this: https://github.com/trinodb/trino/commit/46274c0548af8e2a02ee0534531e5efc176c014e